New Tutorial Track: Typescript with Apollo Server 3 (#1395)
howtographql/howtographql: content/backend/typescript-apollo/1-getting-started.md

---
title: Getting Started
pageTitle: 'Getting Started with GraphQL, Typescript and Apollo'
description:
  'Learn how to setup a GraphQL server with Typescript & Apollo as well as best practices for defining the GraphQL schema.'
question: 'What role do the root fields play for a GraphQL API?'
answers:
  [
    'The three root fields are: Query, Mutation and Subscription',
    'Root fields implement the available API operations',
    'Root fields define the available API operations',
    'Root field is another term for selection set'
  ]
correctAnswer: 2
---

In this section, you will set up the project for your GraphQL server, create your first GraphQL schema and run your first query. In the end,
you will understand some theory behind GraphQL and concepts related to the
[GraphQL schema](https://www.prisma.io/blog/graphql-server-basics-the-schema-ac5e2950214e).

### Checking your node version

> Before you begin, make sure that you are using a suitable version of Node.js. For this tutorial, your Node version should be `v12` or above. 

Reviewer: ```suggestion
Before you begin, make sure that you are using a suitable version of Node.js. For this tutorial, your Node.js version should be `v12` or above. 
```

<Instruction>

> Use the following command to check the version of `node` you are using.

Reviewer: ```suggestion
Use the following command to check the version of `node` you are using:
```

```bash
node --version
```

</Instruction>

> If you need to upgrade your node version, you can do so directly from the [nodejs website](https://nodejs.org/en/) or using a version manager such as [nvm](https://github.com/nvm-sh/nvm). 

Reviewer: ```suggestion
If you need to upgrade your Node.js version, you can do so directly from the [Node.js website](https://nodejs.org/en/) or using a version manager such as [nvm](https://github.com/nvm-sh/nvm). 
```

### Creating the project

This tutorial teaches you how to build a GraphQL server from scratch, so the first thing you need to do is create the
directory that'll hold the files for your GraphQL server!

<Instruction>

Open your terminal, navigate to a location of your choice, and run the following commands:

```bash
mkdir hackernews-typescript
cd hackernews-typescript
npm init -y
```

</Instruction>

This creates a new directory called `hackernews-typescript` and initializes it with a `package.json` file. `package.json` is
the configuration file for the Node.js app you're building. It lists all dependencies and other configuration options
(such as _scripts_) needed for the app.

> ### Installing and configuring typescript

Reviewer: ```suggestion
### Installing and configuring TypeScript
```

Since this is a typescript tutorial, you will need to install `typescript`. You will also install `ts-node-dev`, which will enable you to transpile your TS files on the fly and restart your API on changes. You will soon see that this functionality will come in very handy during development. You will also need a `tsconfig.json` file to specify various Typescript compiler options. 

<Instruction>

> In your terminal, go to the directory you created previously and install `typescript` and `ts-node-dev` as dev dependencies. You will also create a `tsconfig.json` file.

Reviewer: ```suggestion
In your terminal, go to the directory you created previously and install `typescript` and `ts-node-dev` as dev dependencies. You will also create a `tsconfig.json` file:
```

```bash(path=".../hackernews-typescript/")
npm install --save-dev typescript@^4.3.5 ts-node-dev@^1.1.8
touch tsconfig.json              
```

</Instruction>

> **Note**: The above code block is annotated with a directory name. It indicates where you need to execute the terminal command.

Great! Now that you have an empty `tsconfig` file, you need to specify the configuration options. For this project, you will use fairly standard options.

<Instruction> 

> Copy paste the configuration to your `tsconfig.json`

Reviewer: ```suggestion
Copy paste the configuration to your `tsconfig.json`:
```

{
  "compilerOptions": {
    "target": "ES2018",
    "module": "commonjs",
    "lib": [
      "esnext"
    ],
    "strict": true,
    "rootDir": ".",
    "outDir": "dist",
    "sourceMap": true,
    "esModuleInterop": true
  }
}
```
</Instruction>

> **Note:** Setting `"strict": true` enables a wide range of type checking behavior that results in more type-safe programs. However it can also lead to some type errors that are actually rather benign. If you get stuck with a type error that you know isn't a bug, but you are having difficulty satisfying the typescript compiler, you could consider turning this to `false` or using the `// @ts-ignore` comment to turn off that specific error/warning. 

To learn more about the options used here, you can check out the [tsconfig reference](https://www.typescriptlang.org/tsconfig) in the Typescript documentation. 

### Creating your server layout

Before writing any code, you will need to install some dependencies. You will need `nexus` for generating the GraphQL schema. You will also need to install `graphql` and `apollo-server` for creating an HTTP web server with GraphQL capabilities. 

<Instruction>

> Install the necessary packages along with their type definitions

Reviewer: ```suggestion
Install the necessary packages along with their type definitions:
```

```bash
npm install apollo-server@^3.1.1 graphql@^15.5.1 nexus@^1.1.0
```

</Instruction>

[`apollo-server`](https://github.com/apollographql/apollo-server/tree/main/packages/apollo-server) is a fully-featured
GraphQL server. It is based on [Express.js](https://expressjs.com/) and a few other libraries to help you build
production-ready GraphQL servers.

Here's a list of some cool features it has:

- GraphQL spec-compliant
- Out-of-the-box support for two GraphQL clients: Apollo Explorer and GraphQL Playground 

> - Can be configured for use with an [Express](https://expressjs.com/) server. 

Reviewer: ```suggestion
- Can be configured for use with an [Express](https://expressjs.com/) server
```

- Query performance tracing
- Runs everywhere: Can be deployed via Vercel, Standard VMs, AWS Lambda, Heroku etc.

> [`nexus`](https://github.com/graphql-nexus/nexus) is a library to create Type-Safe GraphQL Schemas with a Code-first approach (Since you write standard javascript/typescript code to define what your schema will look like, hence "code-first"). 

Reviewer: ```suggestion
[`nexus`](https://github.com/graphql-nexus/nexus) is a library to create type-safe GraphQL schemas with a [code-first](https://www.prisma.io/blog/the-problems-of-schema-first-graphql-development-x1mn4cb0tyl3) approach (since you write standard javascript/typescript code to define what your schema will look like, hence "code-first"). 
```

It has a number of great features:   

- Expressive, declarative API for building schemas
- Full type-safety for free

> - Auto-generated graphql schema (`schema.graphql` file)

Reviewer: ```suggestion
- Auto-generated GraphQL SDL (`schema.graphql` file)
```

> - Works out of the box with existing graphql frameworks and middleware (`apollo-server`, `graphql-middleware`, etc.) 

Reviewer: ```suggestion
- Works out of the box with existing GraphQL frameworks and middleware (`apollo-server`, `graphql-middleware`, etc.) 
```

> With the project directory in place and dependencies installed, you will now set up a rudimentary GraphQL server. To start off, you will need two files, `schema.ts` for generating the schema with nexus and `index.ts` for creating a GraphQL web server with apollo.

Reviewer: ```suggestion
With the project directory in place and dependencies installed, you will now set up a rudimentary GraphQL server. To start off, you will need two files: 
- `schema.ts` for generating the schema with Nexus
- `index.ts` for creating a GraphQL web server with Apollo
```

<Instruction>

> Create a `src` folder, create the necessary typescript files. 

Reviewer: ```suggestion
Create a `src` folder, create the necessary typescript files:
```

import { makeSchema } from 'nexus'
import { join } from 'path'

export const schema = makeSchema({
  types: [], // 1
  outputs: {
    schema: join(__dirname, '..', 'schema.graphql'), // 2
    typegen: join(__dirname, '..', 'nexus-typegen.ts'), // 3
  },
})
```
</Instruction>

All right, let’s understand what’s going on here by walking through the numbered comments:

1. Your GraphQL schema will consist of many types that you will pass as an array to the `types` object. For now, it is intentionally kept empty.  
2. The first output file that `nexus` will generate for you is a GraphQL `schema` file of type `.graphql`. This is the GraphQL _Schema Definition Language_ (SDL) for defining the structure of your API. You will learn more about this later in the chapter!
3. The second output file is a typescript file known as `typegen`, which will contain typescript type definitions for all types in your GraphQL schema. These generated types will help ensure typesafety in your application code and keep your GraphQL Schema _definition_ in sync with your schema _implementation_. Again, more on this later. 

Let's see what happens when we run this code. 

<Instruction>

> Run the following command in your terminal

Reviewer: ```suggestion
Run the following command in your terminal:
```

```bash(path=".../hackernews-typescript/")
npx ts-node --transpile-only src/schema
```

</Instruction>

After running the script you should see two new files inside your root folder, `schema.graphql` and `nexus-typegen.ts`; these were generated by nexus. 

> The `schema.graphql` file contains a type called `Query` with a single field `ok`. This was created as a default schema, as you did not give nexus any further information about what types you want in the schema.

Reviewer: ```suggestion
The `schema.graphql` file contains a type called `Query` with a single field `ok`. This was created as a default schema, as you did not give Nexus any further information about what types you want in the schema.
```

```graphql(path=".../hackernews-typescript/schema.graphql"&nocopy)
type Query {
  ok: Boolean!
}
```

> The `nexus-typegen.ts` contains a lot of autogenerated typescript `interfaces` and `types`. You won't have to dig too deep into these, as these will be added automatically to your nexus function signatures. Though at certain times, you might look up a type from here and use it manually in your code. 

Reviewer: ```suggestion
The `nexus-typegen.ts` contains a lot of auto-generated TypeScript `interfaces` and `types`. You won't have to dig too deep into these, as these will be added automatically to your nexus function signatures. Though at certain times, you might look up a type from here and use it manually in your code. 
```

Before you move on to the next section, you will add a `generate` script to your `package.json` so you can quickly regenerate your schema at any time. You will also add another `dev` script, which will come in handy in the next section to start our web server. 

<Instruction>

> Add the `generate` and `dev` script to your package.json

Reviewer: ```suggestion
Add the `generate` and `dev` script to your `package.json`:
```

```json{3-4}(path=".../hackernews-typescript/package.json")
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "dev": "ts-node-dev --transpile-only --no-notify --exit-child src/index.ts",
    "generate": "ts-node --transpile-only src/schema.ts"
  },
```
</Instruction>

1. You can run `npm run generate` to update your `schema.graphql` and `nexus-typegen.ts` file when there are any changes in your `nexus` code.
2. You can use `npm run dev` to start the web server and watch for any changes. 

### Creating a GraphQL server

Now that you have a schema, you can finally create your GraphQl server inside `index.ts`. 

<Instruction>

> Write up the following code to start the server in `src/index.ts`. 

Reviewer: ```suggestion
Write up the following code to start the server in `src/index.ts`:
```

```typescript(path=".../hackernews-typescript/src/index.ts")
import { ApolloServer } from "apollo-server";

// 1
import { schema } from "./schema";
export const server = new ApolloServer({
    schema,
});

const port = 3000;
// 2
server.listen({port}).then(({ url }) => {
    console.log(`🚀  Server ready at ${url}`);
});

```

</Instruction>

Let's understand what's going on here by walking through the numbered comments: 

> 1. The schema object you created using `nexus` defines your GraphQL schema. You need to provide this when instantiating your server.

Reviewer: ```suggestion
1. The `schema` object you created using `nexus` defines your GraphQL schema. You need to provide this when instantiating your server.
```

2. You start the server and specify the port. After the server starts, it returns a `url` string inside a promise. 

Now that your server is ready let's give it a run and see what happens!

<Instruction> 

> Open up your terminal and run the following command

Reviewer: ```suggestion
Open up your terminal and run the following command:
```

```bash(path=".../hackernews-typescript/")
npm run dev
```

</Instruction>

If everything goes well, you should see the following output inside the terminal:

```(nocopy)
🚀  Server ready at http://localhost:3000/
```

Now if you navigate to `http://localhost:3000/` you should see the following page: 

![Apollo Server Landing Page](https://i.imgur.com/2bOq8c1.png)

> If you click on the `Query your Server` button, you will be redirected to The Apollo Studio Explorer. This is an _online web-based GraphQL IDE_ for running queries and exploring your schema (along with many other advanced features). 

Reviewer: When referencing UI elements in a text, I've come to adopt a consistent technique of always making the text of the UI element bold.

The rational is that it stands out of the current text and gives a strong visual anchor to the UI element. While I think that with consistent usage, inline code would also work, I prefer using inline code formatting for actual code, file names, CLI commands, ...

If you want to see an example of where I applied these bold visual anchors quite a bit, see this tutorial: https://dev.to/prisma/how-to-setup-a-free-postgresql-database-on-heroku-1dc1#:~:text=Heroku%20offers%20a%20free%20plan,database%22%20for%20experimentation%20or%20prototyping.

```suggestion
If you click on the **Query your server** button, you will be redirected to The Apollo Studio Explorer. This is an _online web-based GraphQL IDE_ for running queries and exploring your schema (along with many other advanced features). 
```

Reviewer: ```suggestion
If you click on the `Query your Server` button, you will be redirected to the [Apollo Studio Explorer](https://www.apollographql.com/docs/studio/explorer/). This is an _online web-based GraphQL IDE_ for running queries and exploring your schema (along with many other advanced features). 
```

> **Note:** If you'd prefer an _offline_ IDE that does not need access to the internet, there's a section covering this at the bottom of this chapter. 

![Apollo Studio Explorer](https://i.imgur.com/jsDqCAi.png)

> In the schema tab placed on the left (the first icon under the apollo logo), you can see the entire GraphQL schema. There's not much there right now, but in the future it will come in handy to explore the details of your schema. 

Reviewer: (In this case the visual anchor doesn't work because "Schema" is not spelled out on the tab, but I'd suggest to still stick to the convention)

```suggestion
In the **Schema** tab placed on the left (the first icon under the apollo logo), you can see the entire GraphQL schema. There's not much there right now, but in the future it will come in handy to explore the details of your schema. 
```

Reviewer: ```suggestion
In the schema tab placed on the left (the first icon under the Apollo logo), you can see the entire GraphQL schema. There's not much there right now, but in the future it will come in handy to explore the details of your schema. 
```

![Schema Apollo Studio](https://i.imgur.com/fJK6FGq.gif)

Now, you will run the `ok` Query.

<Instruction> 

> Go back to the explorer tab (below the Schema tab) and type the following query in the `Operations` window. Then press the `Query` button.

Reviewer: ```suggestion
Go back to the **Explorer** tab (below the **Schema** tab) and type the following query in the **Operations** window. Then press the **Query** button:
```

```graphql
query Query {
  ok
}
```

</Instruction> 

You should see a response like this: 

> ```graphql(nocopy)

Reviewer: ```suggestion
```js(nocopy)
```

{
  "data": {
    "ok": true
  }
}
```

![Run Ok Query](https://i.imgur.com/UVQVOTu.gif)

Congratulations, you just implemented and successfully tested your first GraphQL query 🎉

> > **Note:** Quickly try something, make a quick change *anywhere* in your code. You will see the server restart again with a  message in your console log saying "Restarting: ...". This is because `ts-node-dev` will constantly check your code for any changes and update the server to reflect those changes. **From now on, you will always keep ts-node-dev running.** This will not only ensure that your server is always running, but it will also ensure nexus generates the most updated representation of your GraphQL SDL and types. 

Reviewer: ```suggestion
> **Note:** Quickly try something, make a quick change *anywhere* in your code. You will see the server restart again with a  message in your logging output saying `"Restarting: ..."`. This is because `ts-node-dev` will constantly check your code for any changes and update the server to reflect those changes. **From now on, you will always keep `ts-node-dev running`.** This will not only ensure that your server is always running, but it will also ensure Nexus generates the most updated representation of your GraphQL SDL and types. 
```

Reviewer: I'd either not make this a not since it contains an instruction ("Quickly try something, make a quick change *anywhere* in your code") or make it clearer that this instruction is optional.

### A word on the GraphQL schema

> Now it's time to understand the basics of a GraphQL Schema. At the core of every GraphQL API, there is a GraphQL schema. So, let's quickly talk about it.

Reviewer: ```suggestion
Now it's time to understand the basics of a GraphQL schema. At the core of every GraphQL API, there is a GraphQL schema. So, let's quickly talk about it.
```

  them. Note, however, when querying an object type, it is required that you query at least one of its fields in a
  selection set.
- For the fields in the selection set, it doesn't matter whether the type of the root field is _required_ or a _list_.
  In the example schema above, the three root fields all have different
  [type modifiers](http://graphql.org/learn/schema/#lists-and-non-null) (i.e. different combinations of being a list
  and/or required) for the `User` type:
  - For the `users` field, the return type `[User!]!` means it returns a _list_ (which itself cannot be `null`) of
    `User` elements. The list can also not contain elements that are `null`. So, you're always guaranteed to either
    receive an empty list or a list that only contains non-null `User` objects.
  - For the `user(id: ID!)` field, the return type `User` means the returned value could be `null` _or_ a `User` object.
  - For the `createUser(name: String!)` field, the return type `User!` means this operation always returns a `User`
    object.

Phew, enough theory 😠 Let's go and write some more code!

> ### **OPTIONAL:** GraphQL Playground, An Offline alternative to Apollo Studio Explorer

Reviewer: Bold and uppercase feels like a bit too much here 😅 

```suggestion
### **Optional:** GraphQL Playground, an offline alternative to Apollo Studio Explorer
```

    schema: join(__dirname, '..', 'schema.graphql'), // 2
    typegen: join(__dirname, '..', 'nexus-typegen.ts'), // 3
  },
})
```
</Instruction>

All right, let’s understand what’s going on here by walking through the numbered comments:

1. Your GraphQL schema will consist of many types that you will pass as an array to the `types` object. For now, it is intentionally kept empty.  
2. The first output file that Nexus will generate for you is a GraphQL `schema` file of type `.graphql`. This is the GraphQL _Schema Definition Language_ (SDL) for defining the structure of your API. You will learn more about this later in the chapter!
3. The second output file is a TypeScript file known as `typegen`, which will contain TypeScript type definitions for all types in your GraphQL schema. These generated types will help ensure typesafety in your application code and keep your GraphQL schema _definition_ in sync with your schema _implementation_. Again, more on this later. 

Let's see what happens when we run this code. 

<Instruction>

Run the following command in your terminal:

```bash(path=".../hackernews-typescript/")
npx ts-node --transpile-only src/schema
```

</Instruction>

> After running the script you should see two new files inside your root folder, `schema.graphql` and `nexus-typegen.ts`; these were generated by nexus. 

Reviewer: ```suggestion
After running the script you should see two new files inside your root folder, `schema.graphql` and `nexus-typegen.ts`; these were generated by Nexus. 
```

The `schema.graphql` file contains a type called `Query` with a single field `ok`. This was created as a default schema, as you did not give Nexus any further information about what types you want in the schema.

```graphql(path=".../hackernews-typescript/schema.graphql"&nocopy)
type Query {
  ok: Boolean!
}
```

> The `nexus-typegen.ts` contains a lot of auto-generated TypeScript `interfaces` and `types`. You won't have to dig too deep into these, as these will be added automatically to your nexus function signatures. Though at certain times, you might look up a type from here and use it manually in your code. 

Author: Yeah, that makes sense. Noted!
